### ISSUE TITLE
file_packager.py: option to continue-on-error (for optional packages)?

### ISSUE THREAD

**Issue by past-due** [emscripten-core/emscripten#21310]
Currently, when using `file_packager.py` to generate a preload package*, the output .js code will throw an error inside `fetchRemotePackage` if the associated `.data` file cannot be downloaded. This then yields a state where loading the Module is blocked indefinitely.

<sup>* (via `file_packager.py --preload [...] --js-output=packageName.js --use-preload-cache`)</sup>

In the case of optional packages (which an application can handle running without), we'd like to essentially swallow this error and allow loading / running to continue.

The specific use-case: We have optional packages, and an application we're trying to make resilient in offline scenarios (ex. when the .data files for those packages aren't cached). The application can already handle if the optional packages aren't present, so all that's needed is functionality in the `file_packager` generated script to gracefully handle the error case and remove the run dependencies.

Thoughts on options?

One idea:

Have `file_packager.py` take a new `--continue-on-fetch-error`/`--continue-on-loading-error`/`--continue-on-error` (?) option that adapts the output .js code to:
1. Log any error, but either handle or not throw:
   `throw new Error("NetworkError for: " + packageName);`
   `throw new Error(xhr.statusText + " : " + xhr.responseURL);`
   (which currently end up as unhandled errors)
2. (Optionally) Call a special event on Module that could be implemented for external pre/post-js scripts that want to do custom error handling
3. Remove all run dependencies on error, so that running is unblocked
   `Module['removeRunDependency']('datafile_<packageName>.data');`
   ``Module['removeRunDependency'](`fp ${that.name}`);`` (for all dependencies added as a result of `DataRequest.open` calls in `runWithFS`

**Comment by sbc100** [comment]
If possible, I'd would prefer not to add more feature to `file_package.py` itself.  Would it be possible to take the generated code and somehow modify/wrap it in try/catch to get the behaviour you want?   

If we do end up adding new option perhaps we should just call it `--optional`?

**Comment by past-due** [comment]
> If possible, I'd would prefer not to add more feature to `file_package.py` itself. Would it be possible to take the generated code and somehow modify/wrap it in try/catch to get the behaviour you want?
> 

@sbc100:

One of the issues encountered investigating such an approach is how to handle unblocking the WebAssembly module.

The file packager's generated script adds multiple run dependencies:
1. One run dependency based on the packageName (`datafile_<packageName>.data`) - which is consistent, and could have a corresponding `removeRunDependency` call hard-coded in external error handling.
2. Individual run dependencies (`` `fp ${that.name}` ``) for _every single file that ends up packaged_ (as a result of every `DataRequest.open` call via `runWithFS`).

It appears that 2 is a sticking point (although I may be missing something). And in any event, having to replicate in an external script the internal implementation details of how `file_packager.py` currently names the run dependencies that it adds... doesn't seem ideal (and seems rather brittle).

(Hence why I went with adding an option to `file_packager.py`.)

> If we do end up adding new option perhaps we should just call it `--optional`?

I've gone ahead and updated the PR to name it `--optional`, in case we end up circling back around to this approach.

**Comment by past-due** [comment]
> Regarding (2), are you saying that the`fp ${that.name}` dependencies are per-file? i.e. if you bundle 1000 files it will make 1000 run dependencies?

Yes - for every file, a `DataRequest` is created and `DataRequest.open` is called:
https://github.com/emscripten-core/emscripten/blob/52e5687932555256df6a89d942ffd1c3b90a646b/tools/file_packager.py#L698-L701

And `DataRequest.open` calls `addRunDependency`:
https://github.com/emscripten-core/emscripten/blob/52e5687932555256df6a89d942ffd1c3b90a646b/tools/file_packager.py#L679-L685
